cli/internal/util/backends.go

```diff
@@ -0,0 +1,33 @@
+package util
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/Masterminds/semver"
+)
+
+func IsYarn(backendName string) bool {
+	return backendName == "nodejs-yarn" || backendName == "nodejs-berry"
+}
+
+func IsBerry(cwd string) (bool, error) {
+	cmd := exec.Command("yarn", "--version")
```

Comment:
After bringing this up internally w/Vercel team, we actually think the real way we should do package manager detection is by looking for the recently released/supported `packageManager` key in `package.json`. This will simplify `Detect()` code dramatically.



So let's refactor `api.backend` to look for the package manager key by default now (this will require passing `PackageJson` to the `GetBackend` function and then to respective `Detect()` functions as well. Let's also do this so that we avoid reading `package.json` multiple times. 

Below is a link to relevant Node.js docs and also part of Yarn berry's detection logic. LEt's rip out the highlighted parts into their respective `Detect()` functions. 

**Node.js Documentation:** https://nodejs.org/api/packages.html#determining-package-manager 
**Yarn's Detection Logic** https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-core/sources/scriptUtils.ts#L57-L79


We will ship this as part of `turbo@1.1.x` which will also include a migration script that will set the `packageManager` key for all packages in the workspace. This migration script should be part of a package I started working on called `turbomod` here: https://github.com/vercel/turborepo/pull/218/files#diff-e71d4eb9cf9fafa5c7175aded9c8eebd84ea80cb2e3d6042e3afafbe2f8b772cR1-R23. The script I wrote was to move turbo config to `turbo.json`. I will file an issue with more details on behavior. 

For the purposes of this PR, we should modify the error message we throw when we cannot determine the backend to suggest to user to set the `packageManager` key

Reply:
Using `/(npm|pnpm|yarn)@(\d+)\.\d+\.\d+(-.+)?/`, `$2` will be the major version.

_(Regex found in the `pacakge.json` schema)_
https://github.com/SchemaStore/schemastore/blob/6ba4772130bd07393a8f16f2918663f9be9e0d2c/src/schemas/json/package.json#L561

The yarn code just checks if it is different than 1.

https://github.com/yarnpkg/berry/blob/cb8b6dccf7529385693feded212109455361945e/packages/yarnpkg-core/sources/scriptUtils.ts#L66

Reply:
Do I leave the current implementation as a fallback if the `packageManager` field isn't set until you make the `turbomod` script?

Reply:
Refactor context so that we read the root `package.json` first. Then detect backend. Then later ResolveRoot deps etc.

Reply:
Alright.

Reply:
You rock!

Reply:
Do I leave the current implementation as a fallback if the packageManager field isn't set or just have it error out and tell the user to set it?

Reply:
Was doing some testing and noticed #549 is happening on this version of Turbo too.

Reply:
And now it's working again. I don't know what changed.

Reply:
Remove it